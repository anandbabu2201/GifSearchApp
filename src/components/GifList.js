import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import Gif from './Gif'
import { getGifs,getTrendingGif } from "../actions/gifActions";

class GifList extends Component {

  handleMore = () =>{
    let { total_count, count }= this.props.pageSize || {};
    let { searchBy , filterBy }= this.props.params
    if(filterBy.toUpperCase() === 'RANDOM') return;
    if(count>=total_count) return;
    if(searchBy)
    this.props.getGifs(searchBy,count+25);
    else this.props.getTrendingGif(count+25)
  }

  handleScroll = (e) => {
    if (e.target.className !== 'gifList' || this.props.params.filterBy.toUpperCase() === 'RANDOM' ) return;
    if ((e.target.offsetHeight + e.target.scrollTop) === (e.target.scrollHeight)) this.handleMore();
  }


  render() {
    const { gifData = [] } = this.props;
   return (
      <div className="gifList" onScroll={this.handleScroll} >
        {Array.isArray(gifData)?gifData.map(each => (
          <Gif key={each.id} gif={each} />
        )):<Gif gif={gifData} />}
      </div>
    );
  }
}

GifList.propTypes = {
  gifData: PropTypes.oneOfType([PropTypes.array,PropTypes.object]),
  getGifs: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  gifData: state.gifState.gifData.data,
  pageSize : state.gifState.gifData.pagination,
  params:state.gifState.params
});

export default connect(
  mapStateToProps,
  { getGifs,getTrendingGif }
)(GifList);
