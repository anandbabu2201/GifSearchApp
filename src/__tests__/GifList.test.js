import React from 'react';
import {shallow } from 'enzyme';
import GifList from '../components/GifList';
import thunk from 'redux-thunk';
import configureMockStore from "redux-mock-store";

const mockStore = configureMockStore([ thunk ]);
let store ;
let wrapper;

beforeEach(() => {
  store = mockStore({
    gifState:{
      gifData:{
      data:[
        {
          title:"test",
          "id":1,
          images:
            {
              "480w_still":{url: "sampleimage.jpg"},
              "downsized":{url: "simpletext.gif" }
             },
          },
          {
            title:"test",
            "id":2,
            images:
              {
                "480w_still":{url: "sampleimage.jpg"},
                "downsized":{url: "simpletext.gif" }
               },
            }
      ],
      "pagination":{
        "count":25,
        "total_count":7000

      }
      },
      "params":{
        filterBy:'',
      searchBy:"cat"
      }
    }
  });
  wrapper = shallow(<GifList store={store} />).shallow().dive();
});


describe('Testing GifList component',() => {
  it('renders GifList component', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('it should render two Gifs',()=> {
  expect(wrapper.find('Gif')).toHaveLength(2);
  })
  it ('it should call handlescroll method on scroll of the component',()=>
  {
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:20,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })
  it ('it should call handlescroll but shuould not when filter by random',()=>
  {
    store = mockStore({
      gifState:{
        gifData:{
        data:[
          {
            title:"test",
            "id":1,
            images:
              {
                "480w_still":{url: "sampleimage.jpg"},
                "downsized":{url: "simpletext.gif" }
               },
            },
            {
              title:"test",
              "id":2,
              images:
                {
                  "480w_still":{url: "sampleimage.jpg"},
                  "downsized":{url: "simpletext.gif" }
                 },
              }
        ],
        "pagination":{
          "count":25,
          "total_count":7000

        }
        },
        "params":{
          filterBy:'RANDOM',
        searchBy:"cat"
        }
      }
    });
    wrapper = shallow(<GifList store={store} />).shallow().dive();
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:20,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })

  it ('it should not call getGifs when total count less than count',()=>
  {
    store = mockStore({
      gifState:{
        gifData:{
        data:[
          {
            title:"test",
            "id":1,
            images:
              {
                "480w_still":{url: "sampleimage.jpg"},
                "downsized":{url: "simpletext.gif" }
               },
            },
            {
              title:"test",
              "id":2,
              images:
                {
                  "480w_still":{url: "sampleimage.jpg"},
                  "downsized":{url: "simpletext.gif" }
                 },
              }
        ],
        "pagination":{
          "count":25,
          "total_count":12

        }
        },
        "params":{
          filterBy:'',
        searchBy:"cat"
        }
      }
    });
    wrapper = shallow(<GifList store={store} />).shallow().dive();
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:20,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })

  it ('it should  call getTrendingGifs when filterby trending',()=>
  {
    store = mockStore({
      gifState:{
        gifData:{
        data:[
          {
            title:"test",
            "id":1,
            images:
              {
                "480w_still":{url: "sampleimage.jpg"},
                "downsized":{url: "simpletext.gif" }
               },
            },
            {
              title:"test",
              "id":2,
              images:
                {
                  "480w_still":{url: "sampleimage.jpg"},
                  "downsized":{url: "simpletext.gif" }
                 },
              }
        ],
        "pagination":{
          "count":25,
          "total_count":120

        }
        },
        "params":{
          filterBy:'Trending'
        }
      }
    });
    wrapper = shallow(<GifList store={store} />).shallow().dive();
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:20,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })

  it ('it should call handlescroll but should not go for data call when scroll didn not reach last node',()=>
  {
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:10,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })

  it ('to check whtere gif array or gif object is coming',()=>
  {
    store = mockStore({
      gifState:{
        gifData:{
        data:{
          title:"test",
          "id":1,
          images:
            {
              "480w_still":{url: "sampleimage.jpg"},
              "downsized":{url: "simpletext.gif" }
             },
          },
        "pagination":{
          "count":25,
          "total_count":120

        }
        },
        "params":{
          filterBy:'Trending'
        }
      }
    });
    wrapper = shallow(<GifList store={store} />).shallow().dive();
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:20,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })

  it ('passing default data with',()=>
  {
    store = mockStore({
      gifState:{
        gifData:{
        "pagination":{
          "count":25,
          "total_count":120

        }
        },
        "params":{
          filterBy:'Trending'
        }
      }
    });
    wrapper = shallow(<GifList store={store} />).shallow().dive();
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:20,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })

  it ('passing empty paginzation',()=>
  {
    store = mockStore({
      gifState:{
        gifData:{
        data:[
          {
            title:"test",
            "id":1,
            images:
              {
                "480w_still":{url: "sampleimage.jpg"},
                "downsized":{url: "simpletext.gif" }
               },
            },
            {
              title:"test",
              "id":2,
              images:
                {
                  "480w_still":{url: "sampleimage.jpg"},
                  "downsized":{url: "simpletext.gif" }
                 },
              }
        ],
        },
        "params":{
          filterBy:'Trending'
        }
      }
    });
    wrapper = shallow(<GifList store={store} />).shallow().dive();
    const event = {
      target: { offsetHeight: 10,
        className:'gifList',
        scrollTop:20,
      scrollHeight:30
     }
    };
    wrapper.find('div').simulate('scroll',event);
  })

});
