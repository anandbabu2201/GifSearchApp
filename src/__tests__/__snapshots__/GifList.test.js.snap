// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifList component renders GifList component 1`] = `ShallowWrapper {}`;
